File: DsDotNet/src/Engine/Engine.Parser/3.1.SkeletonListener.cs
using Microsoft.FSharp.Core;

namespace Engine.Parser;


/// <summary>
/// System, Flow, Parenting(껍데기만),
/// Interface name map 구조까지 생성
/// Element path map 구성
///   - Parenting, Child, alias, Api
/// </summary>
class SkeletonListener : ListenerBase
{
    public SkeletonListener(dsParser parser, ParserHelper helper)
        : base(parser, helper)
    {
    }

    override public void EnterSystem(SystemContext ctx)
    {
        if (findFirstChild<SysBlockContext>(ctx) != null)
        {
            var name = ctx.systemName().GetText().DeQuoteOnDemand();
            ICpu cpu = null;    // todo
            var host = findFirstChild<HostContext>(ctx)?.GetText();
            _system = DsSystem.Create(name, host, FSharpOption<ICpu>.None, _model);
            Trace.WriteLine($"System: {name}");
            AddElement(CurrentPathElements, GraphVertexType.System);

        }
    }

    override public void EnterFlow(FlowContext ctx)
    {
        var flowName = ctx.identifier1().GetText().DeQuoteOnDemand();
        _rootFlow = Flow.Create(flowName, _system);
        AddElement(CurrentPathElements, GraphVertexType.Flow);
    }

    override public void EnterParenting(ParentingContext ctx)
    {
        Trace.WriteLine($"Parenting: {ctx.GetText()}");
        var name = ctx.identifier1().GetText().DeQuoteOnDemand();
        _parenting = Segment.Create(name, _rootFlow);
        AddElement(CurrentPathElements, GraphVertexType.Segment | GraphVertexType.Parenting);

        var children =
            enumerateChildren<CausalTokenContext>(ctx)
                .Select(ctctx => collectNameComponents(ctctx).ToArray())
                .Pipe(childNameComponts => Assert(childNameComponts.Length.IsOneOf(1, 2)))
                .Select(childNameComponts => AppendPathElement(childNameComponts))
                .ToArray()
                ;
        foreach(var ch in children)
            AddElement(ch, GraphVertexType.Child);
    }

    override public void EnterIdentifier12Listing(Identifier12ListingContext ctx)
    {
        var ns = AppendPathElement(collectNameComponents(ctx));
        AddElement(ns, GraphVertexType.Segment);
    }

    override public void EnterCausalToken(CausalTokenContext ctx)
    {
        var path = AppendPathElement(collectNameComponents(ctx));
        var vType = GraphVertexType.Call;

        // 다음 stage 에서 처리...
        //if (_parenting == null)
        //    vType |= GraphVertexType.Segment;

        AddElement(path, vType);
    }


    public override void EnterAliasListing(AliasListingContext ctx)
    {
        var map = _rootFlow.AliasMap;

        var aliasDef = findFirstChild<AliasDefContext>(ctx);
        var alias = collectNameComponents(aliasDef);
        switch(alias.Length)
        {
            case 2: // {타시스템}.{interface명} or
                AddElement(alias, GraphVertexType.AliaseKey);
                break;
            case 1: // { (my system / flow /) segment 명 }
                AddElement(AppendPathElement(alias[0]), GraphVertexType.AliaseKey);
                break;
            default:
                throw new Exception("ERROR");
        }

        var mnemonics =
            enumerateChildren<AliasMnemonicContext>(ctx)
                .Select(mctx => collectNameComponents(mctx))
                .Pipe(mne => Assert(mne.Length == 1))
                .Select(mne => mne[0])
                .ToHashSet();
        map.Add(alias, mnemonics);
        foreach(var mne in mnemonics)
            AddElement(AppendPathElement(mne), GraphVertexType.AliaseMnemonic);
    }


    public override void EnterInterfaces([NotNull] InterfacesContext ctx)
    {
        _system.Api = new Api(_system);
    }


    public override void EnterInterfaceDef([NotNull] InterfaceDefContext ctx)
    {
        var hash = _system.Api.Items;
        var interrfaceNameCtx = findFirstChild<InterfaceNameContext>(ctx);
        var interfaceName = collectNameComponents(interrfaceNameCtx)[0];
        string[][] collectCallComponents(CallComponentsContext ctx) =>
            enumerateChildren<Identifier123Context>(ctx)
                .Select(collectNameComponents)
                .ToArray()
                ;
        //string[][] collectCallComponents(CallComponentsContext ctx)
        //{
        //    var xx = enumerateChildren<Identifier123Context>(ctx).ToArray();
        //    var yy = xx.Select(collectNameComponents)
        //        .ToArray()
        //        ;
        //    return yy;

        //}

        //var xxxser =
        //    enumerateChildren<CallComponentsContext>(ctx).ToArray();
        //var zxxxser =
        //    collectCallComponents(xxxser[0]).ToArray();
        //var xxser =
        //    enumerateChildren<CallComponentsContext>(ctx)
        //    .Select(collectCallComponents).ToArray();

        var ser =   // { start ~ end ~ reset }
            enumerateChildren<CallComponentsContext>(ctx)
            .Select(collectCallComponents)
            .Pipe(callComponents => Assert(callComponents.ForAll(cc => cc.Length == 2 || cc[0] == "_")))
            .Select(callCompnents => callCompnents.Select(cc => cc.Prepend(_system.Name).ToArray()).ToArray())
            .ToArray()
            ;

        AddElement(AppendPathElement(interfaceName), GraphVertexType.ApiKey);
        foreach(var cc in ser.SelectMany(x => x))
            AddElement(cc, GraphVertexType.ApiSER);


        // 이번 stage 에서 일단 interface 이름만 이용해서 빈 interface 객체를 생성하고,
        // TXs, RXs, Resets 은 다음 listener stage 에서 채움..
        var api = ApiItem.Create(interfaceName, _system);
        hash.Add(api);
    }

    public override void EnterInterfaceResetDef([NotNull] InterfaceResetDefContext ctx)
    {
        var operands =
            enumerateChildren<Identifier1Context>(ctx)
            .Select(ctx => ctx.GetText())
            .ToArray();
        var operator_ = findFirstChild<CausalOperatorResetContext>(ctx).GetText();
        var ri_ = ApiResetInfo.Create(_system, operands[0], operator_, operands[1]);
    }
}

File: DsDotNet/src/Engine/Engine.Parser.FS/3.1.SkeletonListener.fs
namespace Engine.Parser.FS

open System.Linq
open System.IO

open Antlr4.Runtime.Tree
open Antlr4.Runtime

open Engine.Common.FS
open Engine.Parser
open Engine.Core
open type Engine.Parser.dsParser
open type DsParser


/// <summary>
/// System, Flow, Parenting(껍데기만),
/// Interface name map 구조까지 생성
/// Element path map 구성
///   - Parenting, Child, alias, Api
/// </summary>
type SkeletonListener(parser:dsParser, helper:ParserHelper) =
    inherit ListenerBase(parser, helper)


    override x.EnterSystem(ctx:SystemContext) =
        base.EnterSystem(ctx)

        match tryFindFirstChild<SysBlockContext>(ctx) with
        | Some sysBlockCtx_ ->
            let name = helper.ParserOptions.LoadedSystemName |? (ctx.systemName().GetText().DeQuoteOnDemand())
            let host =
                match tryFindFirstChild<HostContext>(ctx) with
                | Some hostCtx -> hostCtx.GetText()
                | None -> null
            //let name = helper.ParserOptions.LoadedSystemName
            helper.TheSystem <- DsSystem.Create(name, host)
            tracefn($"System: {name}")
        | None ->
            failwith "ERROR"

    override x.EnterFlowBlock(ctx:FlowBlockContext) =
        let flowName = ctx.identifier1().GetText().DeQuoteOnDemand()
        helper._flow <- Flow.Create(flowName, helper.TheSystem)

    override x.EnterParentingBlock(ctx:ParentingBlockContext) =
        helper._parentingBlockContexts.Add(ctx)
        tracefn($"Parenting: {ctx.GetText()}")
        let name = tryGetName(ctx.identifier1()).Value
        helper._parenting <- Real.Create(name, helper._flow)



    override x.EnterCausalPhrase(ctx:CausalPhraseContext) =
        helper._causalPhraseContexts.Add(ctx)

    override x.EnterIdentifier12Listing(ctx:Identifier12ListingContext) =
        helper._identifier12ListingContexts.Add(ctx)

    override x.EnterCausalToken(ctx:CausalTokenContext) =
        helper._causalTokenContext.Add(ctx)




    override x.EnterInterfaceDef(ctx:InterfaceDefContext) =
        helper._interfaceDefContexts.Add(ctx)

        let system = helper.TheSystem
        let interrfaceNameCtx = tryFindFirstChild<InterfaceNameContext>(ctx).Value
        let interfaceName = collectNameComponents(interrfaceNameCtx)[0]

        // 이번 stage 에서 일단 interface 이름만 이용해서 빈 interface 객체를 생성하고,
        // TXs, RXs, Resets 은 추후에 채움..
        let api = ApiItem4Export.Create(interfaceName, system)
        let hash = system.ApiItems4Export
        hash.Add(api) |> ignore

    override x.EnterInterfaceResetDef(ctx:InterfaceResetDefContext) =
        // I1 <||> I2 <||> I3;  ==> [| I1; <||>; I2; <||>; I3; |]
        let terms =
            let pred = fun (tree:IParseTree) -> tree :? Identifier1Context || tree :? CausalOperatorResetContext
            enumerateChildren<RuleContext>(ctx, false, pred)
                .Select(fun ctx -> ctx.GetText())
                .ToArray()

        // I1 <||> I2 와 I2 <||> I3 에 대해서 해석
        for triple in (terms |> Array.windowed2 3 2) do
            if triple.Length = 3 then
                let opnd1, op, opnd2 = triple[0], triple[1], triple[2]
                let ri_ = ApiResetInfo.Create(helper.TheSystem, opnd1, op.ToModelEdge(), opnd2)
                ()

    member private x.GetFilePath(fileSpecCtx:FileSpecContext) =
        let simpleFilePath = tryFindFirstChild<FilePathContext>(fileSpecCtx).Value.GetText().DeQuoteOnDemand()
        let absoluteFilePath =
            let dir = helper.ParserOptions.ReferencePath
            [simpleFilePath; $"{dir}\\{simpleFilePath}"].First(fun f -> File.Exists(f))
        absoluteFilePath, simpleFilePath


    override x.EnterLoadDeviceBlock(ctx:LoadDeviceBlockContext) =
        helper._deviceBlockContexts.Add(ctx)
        let fileSpecCtx = tryFindFirstChild<FileSpecContext>(ctx).Value
        let absoluteFilePath, simpleFilePath = x.GetFilePath(fileSpecCtx)
        let device =
            let loadedName = collectNameComponents(ctx).Combine()
            fwdLoadDevice helper.TheSystem (absoluteFilePath, simpleFilePath) loadedName
        helper.TheSystem.Devices.Add(device) |> ignore

    override x.EnterLoadExternalSystemBlock(ctx:LoadExternalSystemBlockContext) =
        helper._externalSystemBlockContexts.Add(ctx)
        let fileSpecCtx = tryFindFirstChild<FileSpecContext>(ctx).Value
        let absoluteFilePath, simpleFilePath = x.GetFilePath(fileSpecCtx)
        let external =
            let ipSpecCtx = tryFindFirstChild<IpSpecContext>(ctx).Value
            let ip = tryFindFirstChild<EtcNameContext>(ipSpecCtx).Value.GetText()
            let loadedName = collectNameComponents(ctx).Combine()
            fwdLoadExternalSystem helper.TheSystem (absoluteFilePath, simpleFilePath) loadedName
        helper.TheSystem.Devices.Add(external) |> ignore

    override x.EnterAliasListing(ctx:AliasListingContext) =
        helper._aliasListingContexts.Add(ctx)

    override x.EnterCallListing(ctx:CallListingContext) =
        helper._callListingContexts.Add(ctx)

    override x.ExitSystem(ctx:SystemContext) =
        base.ExitSystem(ctx)
        let system = helper.TheSystem

        let getContainerChildPair(ctx:ParserRuleContext) : ParentWrapper option * NamedContextInformation =
            let ci = getContextInformation ctx
            let system = helper.TheSystem
            let parentWrapper = tryFindParentWrapper system ci
            parentWrapper, ci

        let tokenCreator (cycle:int) =
            let candidateCtxs:ParserRuleContext list = [
                yield! helper._identifier12ListingContexts.Cast<ParserRuleContext>()
                yield! helper._causalTokenContext.Cast<ParserRuleContext>()
            ]

            let isCallName (parentWrapper:ParentWrapper) name = tryFindCall system name |> Option.isSome
            let isAliasMnemonic (parentWrapper:ParentWrapper) name =
                let flow = parentWrapper.GetFlow()
                tryFindAliasDefWithMnemonic flow name |> Option.isSome
            let isCallOrAlias pw name =
                if name = "Ap1" then
                    noop()
                let xxx = isCallName pw name
                let yyy = isAliasMnemonic pw name
                isCallName pw name || isAliasMnemonic pw name

            let tryCreateCallOrAlias (parentWrapper:ParentWrapper) name =
                let flow = parentWrapper.GetFlow()
                let tryCall = tryFindCall system name
                let tryAliasDef = tryFindAliasDefWithMnemonic flow name
                option {
                    match tryCall, tryAliasDef with
                    | Some call, None ->
                        return VertexCall.Create(name, call, parentWrapper) :> Indirect
                    | None, Some aliasDef ->
                        let aliasTarget = tryFindAliasTarget flow name |> Option.get
                        return VertexAlias.Create(name, aliasTarget, parentWrapper) :> Indirect
                    | None, None -> return! None
                    | _ ->
                        failwith "ERROR: duplicated"
                }
            let candidates = candidateCtxs.Select(getContainerChildPair)

            let loop () =
                for (optParent, ctxInfo) in candidates do
                    let parent = optParent.Value
                    let existing = parent.GetGraph().TryFindVertex(ctxInfo.GetRawName())
                    match existing with
                    | Some v -> tracefn $"{v.Name} already exists.  Skip creating it."
                    | None ->
                        match cycle, ctxInfo.Names with
                        | 0, q::[] when not <| isCallOrAlias parent q ->
                            let flow = parent.GetCore() :?> Flow
                            Real.Create(q, flow) |> ignore

                        | 1, q::[] when (isCallOrAlias parent q) ->
                            match tryCreateCallOrAlias parent q with
                            | Some _ -> ()
                            | None ->
                                failwith "ERROR"
                        | 1, ofn::ofrn::[] ->
                            let otherFlowReal = tryFindReal system ofn ofrn |> Option.get
                            VertexOtherFlowRealCall.Create(ofn, ofrn, otherFlowReal, parent) |> ignore

                            tracefn $"{ofn}.{ofrn} should already have been created."
                        | _, q::[] -> ()
                        | _, ofn::ofrn::[] -> ()
                        | _ ->
                            failwith "ERROR"
                        noop()
            loop

        let createNonParentedRealVertex = tokenCreator 0
        let createCallOrAliasVertex = tokenCreator 1


        //let dumpTokens (tokens:Dictionary<ContextInformation, GVT>) (msg:string) =
        //    logInfo "%s" msg
        //    for KeyValue(ctxInfo, vType) in tokens do
        //        logDebug $"{ctxInfo.FullName} : {vType}"
        //let dumpCausalTokens = dumpTokens helper._causalTokenElements

        //dumpCausalTokens "---- Original Causal token elements"

        createCallDefs helper
        createAliasDefs helper
        createNonParentedRealVertex()
        fillAliasDefsTarget helper
        createCallOrAliasVertex()
        fillInterfaceDefs helper

        guardedValidateSystem system

        //dumpCausalTokens "---- All Causal token elements"

